File: aaw-backend/src/main/kotlin/com/berno/aaw/entity/Task.kt
package com.berno.aaw.entity

import jakarta.persistence.*
import java.time.LocalDateTime

@Entity
@Table(name = "tasks")
data class Task(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long = 0,

    @Column(nullable = false)
    val instruction: String,

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    var status: TaskStatus = TaskStatus.PENDING,

    @Column(name = "current_branch")
    val currentBranch: String? = null,

    @Column(name = "base_branch")
    val baseBranch: String? = null,

    @Column(name = "jira_key")
    val jiraKey: String? = null,

    @Column(name = "script_content", columnDefinition = "TEXT")
    val scriptContent: String? = null,

    @Column(name = "skip_permissions", nullable = false)
    val skipPermissions: Boolean = false,

    @Enumerated(EnumType.STRING)
    @Column(name = "session_mode", nullable = false)
    val sessionMode: SessionMode = SessionMode.PERSIST,

    @Column(name = "created_at", nullable = false)
    val createdAt: LocalDateTime = LocalDateTime.now(),

    // Phase 4.1: Queue management fields
    @Column(nullable = false)
    val priority: Int = 0,

    @Column(name = "queued_at")
    var queuedAt: LocalDateTime? = null,

    @Column(name = "started_at")
    var startedAt: LocalDateTime? = null,

    @Column(name = "completed_at")
    var completedAt: LocalDateTime? = null,

    @Column(name = "error_summary", length = 1000)
    var errorSummary: String? = null,

    @Enumerated(EnumType.STRING)
    @Column(name = "execution_mode", nullable = false, length = 20)
    val executionMode: ExecutionMode = ExecutionMode.QUEUED,

    @Column(name = "failure_reason", length = 50)
    var failureReason: String? = null,

    @Column(name = "retry_count", nullable = false)
    var retryCount: Int = 0,

    @Column(name = "cancelled_at")
    var cancelledAt: LocalDateTime? = null
)

enum class TaskStatus {
    PENDING,
    QUEUED,
    RUNNING,
    PAUSED,
    RATE_LIMITED,
    COMPLETED,
    FAILED,
    PAUSED_BY_LIMIT,
    INTERRUPTED,
    CANCELLED,
    CANCELLING,  // SIGTERM sent, waiting for graceful shutdown
    KILLED       // Force killed with SIGKILL
}

enum class ExecutionMode {
    QUEUED,
    DIRECT
}

enum class SessionMode {
    NEW,
    PERSIST
}

File: aaw-frontend/types/task.ts
export type TaskStatus =
  | 'PENDING'
  | 'QUEUED'
  | 'RUNNING'
  | 'COMPLETED'
  | 'FAILED'
  | 'PAUSED'
  | 'PAUSED_BY_LIMIT'
  | 'RATE_LIMITED'
  | 'INTERRUPTED'
  | 'CANCELLED'
  | 'CANCELLING'
  | 'TERMINATING'
  | 'KILLED';

export interface RecoveryAttempt {
  attemptNumber: number;
  timestamp: string;
  action: 'RETRY' | 'SKIP' | 'RESTART_SESSION';
  result: 'SUCCESS' | 'FAILED';
  errorMessage?: string;
}

export interface Task {
  id: number;
  instruction: string;
  scriptContent: string;
  scriptPath: string | null;
  status: TaskStatus;
  priority: number;
  queuePosition: number | null;
  skipPermissions: boolean;
  sessionMode: 'NEW' | 'PERSIST';
  executionMode?: 'QUEUED' | 'DIRECT';
  failureReason: string | null;
  retryCount: number;
  createdAt: string;
  startedAt: string | null;
  completedAt: string | null;
  isArchived: boolean;
  deletedAt: string | null;
  recoveryHistory?: RecoveryAttempt[];
}

export interface TaskEvent {
  type: 'TASK_QUEUED' | 'TASK_DEQUEUED' | 'STATUS_UPDATE' | 'TASK_INTERRUPTED' | 'TASK_RUNNING' | 'LOG' | 'SYSTEM' | 'SYSTEM_READY' | 'SYSTEM_DISCONNECTED';
  taskId?: number;
  task?: Task;
  status?: TaskStatus;
  line?: string | null;
  isError?: boolean;
}

/**
 * Phase 4: System state snapshot for frontend re-attachment.
 */
export interface SystemState {
  isRunnerConnected: boolean;
  isRateLimited: boolean;
  runnerStatus: 'IDLE' | 'BUSY';
  runningTasks: RunningTaskInfo[];
  queuedTaskCount: number;
  timestamp: number;
}

export interface RunningTaskInfo {
  id: number;
  instruction: string;
  status: string;
  startedAt: string | null;
}
